sem/utils.py

```diff
@@ -481,3 +489,231 @@ def convert_environment_str_to_dict(log_components):
+                             'Context',
+                             'Component',
+                             'Function',
+                             'Arguments',
+                             'Level',
+                             'Message']
+    }
+
+    for log in logs:
+        # Verify log format is correct
+        # Only check the if the keys are consistent
+        if not(set(log.keys()) == set(example_result.keys())):
+            raise ValueError(
+                '%s:\nExpected: %s\nGot: %s' % (
+                    "Log dictionary does not correspond to database format",
+                    pformat(example_result, depth=2),
+                    pformat(log, depth=2)))
+
+    db.table('logs').insert_multiple(logs)
+
+
+def filter_logs(db,
+                context=[],
+                function=[],
+                time_begin=-1,
+                time_end=-1,
+                level=[],
+                components={}):
+    """
+    Filter the logs stored in the database.
+
+    Filters are applied on context, function name, log level and time.
+    Additionally the user can also filter each log component based on a
+    particular level using components dictionary.
+    For example, if the user specifies Context = [0, 1] and Function = [A, B]
+    the function will output logs in which (context == 0 or context == 1) and
+    (function == a or function == b).
+
+    Return a list of logs that satisfy all the passed filters.
+
+    Args:
+        db (TinyDB instance): A TinyDB instace where the logs are inserted.
+        context (list): A list of context based on which the logs will be
+            filtered.
+        function (list): A list of function names based on which the logs will
+            be filtered.
+        time_begin (float): Start timestamp (in seconds) of the time window.
+        time_end (float): End timestamp (in seconds) of the time window.
+        level (list): A list of log levels based on which the logs will be
+            filtered.
+        components (dict): A dictionary having structure
+            {
+                components:['level1','level2']
+            }
+            based on which the logs will be filtered.
+    """
+    query_final = []
+
+    if level != [] or components != {}:
+        if not isinstance(level, list):
+            level = [level]
+        for value in components.values():
+            if not isinstance(value, list):
+                value = [value]
+
+        query_list = []
+        if level != []:
+            query = reduce(or_,
+                           [where('Level') == lvl.upper() for lvl in level]
+                           )
+            query_list.append(query)
+        # If components is provided apply the specified log levels to the
+        # specified log components in addition to the log levels passed with
+        # 'levels'. In other words, log levels passed with 'levels' is treated
+        # as a global level filter.
```

Comment:
This should be very well documented in the docstring: it is confusing to effectively provide two filter knobs that deal with logging levels/severity classes. We should state clearly that `level` is a global filter, and `components` can be used in addition to that. Even then, it might be better to provide some examples, since the interaction might not be easy to understand: what happens if I define `level_debug` (or `[debug, warn, error]`) as a `level` and then ask for `components: {'componentA': ['debug']}`? Will I see only the `debug` class for `componentA`? Or will I see `debug`, `warn`, and `error`?